ui/src/Websocket.js

```diff
@@ -0,0 +1,54 @@
+var wss = new Object();
+
+MF.log = function(msg) {
+    console.log(msg);
+    app.ports.websocketState.send(msg);
+}
+
+app.ports.connectWebsocket.subscribe(function(data) {
+  var url = 'wss://localhost/ws/channels/' + data.channelid + '/messages?authorization=' + data.thingkey
+  if (wss[url]) {
+    MF.log("Websocket already open. URL: " + url );
+    return;
+  }
+
+  var ws = new WebSocket(url);
+  
+  ws.onopen = function (event) {
+    MF.log("Websocket opened. URL: " + url);
+    wss[url] = ws;          
+  }
+
+  ws.onerror = function (event) {
+    console.log(event);
+  }
+  
+  ws.onmessage = function(message) {
+    app.ports.websocketIn.send(JSON.stringify({data: message.data, timestamp: message.timeStamp}));
+  };
+  
+  ws.onclose = function (event) {
+    MF.log("Websocket closed. URL: " + url);
+    delete wss[ws.url];
+  };
+
+});
+
+app.ports.websocketOut.subscribe(function(data) {
+  var url = 'wss://localhost/ws/channels/' + data.channelid + '/messages?authorization=' + data.thingkey
+  if (wss[url]) {
+    wss[url].send(data.message);
+  } else {
+    MF.log("Websocket is not open. URL: " + url);
+  }
+});
+
+app.ports.disconnectWebsocket.subscribe(function(data) {
+  var url = 'wss://localhost/ws/channels/' + data.channelid + '/messages?authorization=' + data.thingkey
+  if (wss[url]) {
+    wss[url].close();
+  } else {
+    MF.log("Websocket is not open. URL: " + url);
```

Comment:
Is this comment OK? Looks like a copy of precious one.

Reply:
No, it just informs the user that the websocket he's trying to close is not open, so he can't close it. I can reword the text so that it becomes more clear.